includes/documents/class-wcpdf-invoice.php

```diff
@@ -75,34 +92,49 @@ public function exists() {
 	}
 
 	public function init_number() {
-		global $wpdb;
-		// If a third-party plugin claims to generate invoice numbers, trigger this instead
-		if ( apply_filters( 'woocommerce_invoice_number_by_plugin', false ) || apply_filters( 'wpo_wcpdf_external_invoice_number_enabled', false, $this ) ) {
-			$invoice_number = apply_filters( 'woocommerce_generate_invoice_number', null, $this->order );
-			$invoice_number = apply_filters( 'wpo_wcpdf_external_invoice_number', $invoice_number, $this );
-		} elseif ( isset( $this->settings['display_number'] ) && $this->settings['display_number'] == 'order_number' && !empty( $this->order ) ) {
-			$invoice_number = $this->order->get_order_number();
-		}
+		$lock           = new Semaphore( $this->lock_name, $this->lock_time, array( wc_get_logger() ), $this->lock_context );
+		$invoice_number = null;
+		
+		if ( $lock->lock( $this->lock_retries ) ) {
+			
+			try {
+				// If a third-party plugin claims to generate invoice numbers, trigger this instead
+				if ( apply_filters( 'woocommerce_invoice_number_by_plugin', false ) || apply_filters( 'wpo_wcpdf_external_invoice_number_enabled', false, $this ) ) {
+					$invoice_number = apply_filters( 'woocommerce_generate_invoice_number', null, $this->order );
+					$invoice_number = apply_filters( 'wpo_wcpdf_external_invoice_number', $invoice_number, $this );
+				} elseif ( isset( $this->settings['display_number'] ) && $this->settings['display_number'] == 'order_number' && ! empty( $this->order ) ) {
+					$invoice_number = $this->order->get_order_number();
+				}
 
-		if ( !empty( $invoice_number ) ) { // overriden by plugin or set to order number
-			if ( is_numeric($invoice_number) || $invoice_number instanceof Document_Number ) {
-				$this->set_number( $invoice_number );
-			} else {
-				// invoice number is not numeric, treat as formatted
-				// try to extract meaningful number data
-				$formatted_number = $invoice_number;
-				$number = (int) preg_replace('/\D/', '', $invoice_number);
-				$invoice_number = compact( 'number', 'formatted_number' );
-				$this->set_number( $invoice_number );
+				if ( ! empty( $invoice_number ) ) { // overriden by plugin or set to order number
+					if ( ! is_numeric( $invoice_number ) || ! $invoice_number instanceof Document_Number ) {
```

Comment:
I think the way this is structured now might cause issues when an a `Document_Number` object is passed here.

Reply:
I don't think that's a problem because if you look at the old code structure it just required this: `$this->set_number( $invoice_number );` which is done only line `123`.

Reply:
The old code checked the following before moving to `set_number()`:

```
if ( !empty( $invoice_number ) ) { // overriden by plugin or set to order number
	if ( is_numeric($invoice_number) || $invoice_number instanceof Document_Number ) {
		$this->set_number( $invoice_number );
	} else {
		// invoice number is not numeric, treat as formatted
		// try to extract meaningful number data
		$formatted_number = $invoice_number;
		$number = (int) preg_replace('/\D/', '', $invoice_number);
		$invoice_number = compact( 'number', 'formatted_number' );
		$this->set_number( $invoice_number );
	}
	return $invoice_number;
}

$number_store   = $this->get_sequential_number_store();
$invoice_number = $number_store->increment( intval( $this->order_id ), $this->get_date()->date_i18n( 'Y-m-d H:i:s' ) );

$this->set_number( $invoice_number );
```

Which is now replaced by:

```
if ( ! empty( $invoice_number ) ) { // overriden by plugin or set to order number
	if ( ! is_numeric( $invoice_number ) || ! $invoice_number instanceof Document_Number ) {
		// invoice number is not numeric, treat as formatted
		// try to extract meaningful number data
		$formatted_number = $invoice_number;
		$number           = (int) preg_replace( '/\D/', '', $invoice_number );
		$invoice_number   = compact( 'number', 'formatted_number' );
	}
} else {
	$number_store   = $this->get_sequential_number_store();
	$invoice_number = $number_store->increment( intval( $this->order_id ), $this->get_date()->date_i18n( 'Y-m-d H:i:s' ) );
}

if ( ! is_null( $invoice_number ) ) {
	$this->set_number( $invoice_number );
}
```

Note that
```
if ( ! is_numeric( $invoice_number ) || ! $invoice_number instanceof Document_Number )
``` 
does not give you the same result as the `else` statement in the old code:

```
if ( is_numeric($invoice_number) || $invoice_number instanceof Document_Number ) {
        //...
} else {
        //...
}
```

For instance if I use the `wpo_wcpdf_external_invoice_number` filter to generate an invoice number with custom prefix/suffix/padding and return an instance of `Document_Number`,  the check in the new code will result in `true` as `$invoice_number` is not numeric but an object. This then stores the whole `Document_Number` object under `formatted_number` in the order meta. And leaving `prefix`, `suffix` and `padding` blank. 

<img width="483" alt="Schermafbeelding 2023-05-05 om 13 39 47" src="https://user-images.githubusercontent.com/19350150/236452903-a3a94865-f018-4786-881b-9fd3ee9cc313.png">

Reply:
Could you here your code snippet for the `wpo_wcpdf_external_invoice_number` please?

Reply:
I'm still not getting your point of view, maybe I'm missing something here, but let me explain mine:

- If `$invoice_number` is `Document_Number` object:

## Old code

- It goes through this line because it's not empty: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/7dfe79784913dd1a624dab5c4dc15741903dd4db/includes/documents/class-wcpdf-invoice.php#L87
- It goes through this line because it's an object instance of Document_number: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/7dfe79784913dd1a624dab5c4dc15741903dd4db/includes/documents/class-wcpdf-invoice.php#L88
- Number is `set` here: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/7dfe79784913dd1a624dab5c4dc15741903dd4db/includes/documents/class-wcpdf-invoice.php#L89
- Variable is returned after the number is set here: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/7dfe79784913dd1a624dab5c4dc15741903dd4db/includes/documents/class-wcpdf-invoice.php#L98
  
## New code

- It goes through this line because it's not empty: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/f88ffebef964d4dfd374c980f23267fd8c163308/includes/documents/class-wcpdf-invoice.php#L109
- Doesn't go through he next check because it's `Document_number` instance: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/f88ffebef964d4dfd374c980f23267fd8c163308/includes/documents/class-wcpdf-invoice.php#L110
- It goes through this line because it's not null: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/f88ffebef964d4dfd374c980f23267fd8c163308/includes/documents/class-wcpdf-invoice.php#L122
- Sets the number: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/f88ffebef964d4dfd374c980f23267fd8c163308/includes/documents/class-wcpdf-invoice.php#L123
- Variable is returned after the number is set: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/f88ffebef964d4dfd374c980f23267fd8c163308/includes/documents/class-wcpdf-invoice.php#L138

Reply:
> - Doesn't go through the next check because it's `Document_number` instance: https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/f88ffebef964d4dfd374c980f23267fd8c163308/includes/documents/class-wcpdf-invoice.php#L110

It does not go through that check because it isn't numeric. That is why I said:

> Note that
```
if ( ! is_numeric( $invoice_number ) || ! $invoice_number instanceof Document_Number ) {
        //...
}
``` 
does not give you the same result as the `else` statement in the old code:

```
if ( is_numeric($invoice_number) || $invoice_number instanceof Document_Number ) {
        //...
} else {
        //...
}
```

Reply:
You're right, for some reason my head was misleading by that numeric check.